include/drivers/rtio.h

```diff
@@ -0,0 +1,486 @@
+/*
+ * Copyright (c) 2019  Thomas Burdick <thomas.burdick@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_RTIO_H_
+#define ZEPHYR_INCLUDE_DRIVERS_RTIO_H_
+/**
+ * @brief RTIO Interface
+ * @defgroup rtio_interface RTIO Interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#include <kernel.h>
+#include <device.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @file
+ * @brief Real-Time IO device API for moving bytes around fast
+ *
+ * Each Real-Time IO device provides a way to either read, write, or both
+ * blocks of memory allocated from a static pool by the application. The layout
+ * of the each block is understood implicitly by the author. Each device may
+ * provide functionality to safely encode or decode blocks of memory.
+ */
+
+/**
+ * @brief Block of memory
+ *
+ * This block of memory has important metadata commonly useful such as
+ * byte layout, and timestamps of when a read or write began and ended.
+ */
+struct rtio_block {
+	/** Timestamp in cycles from k_cycle_get_32() marking when a read or
+	 * write began
+	 */
+	u32_t begin_tstamp;
+
+	/** Timestamp in cycles from k_cycle_get_32() marking when a read or
+	 * write was complete
+	 */
+	u32_t end_tstamp;
+
+	/** Byte layout designator given by each driver to specify
+	 * the layout of the contained buffer
+	 */
+	u32_t layout;
+
+	/** Length of data in this buffer. */
+	size_t len;
+
+	/** Amount of data this buffer can store. */
+	size_t size;
+
+	/** Start of the buffer data .*/
+	u8_t *_data;
```

Comment:
If this is always a pointer to the end of the structure, never to an external buffer, maybe it would make sense to have it be something like:

    union {
       u8_t u8[1];
       u32_t alignment; /* or whatever alignment restriction is needed for the data */
    } data;

to avoid the wasted space for a pointer and the code to dereference it.  Would add a slight complication to calculating the actual allocation size.

Reply:
Its not unfortunately. The way I've allowed for different allocators so far is by creating things like 

``` struct rtio_mempool_block {
  struct rtio_block block;
  struct k_mempool_id mempool_id;
} ```

If there's a better way to do this sort of thing in C I'm open to suggestions.